### ISSUE TITLE
armv7a: Error: co-processor offset out of range in kernel_dgemm_4x4_lib4.S

### ISSUE THREAD

**Issue by stefanct** [giaf/blasfeo#108]
When compiling with `TARGET_ARMV7A_ARM_CORTEX_A15` (and at least `A7`) with the current gcc in Debian stable (Buster) which is arm-linux-gnueabihf-gcc (Debian 8.3.0-2) 8.3.0 `kernel/armv7a/kernel_dgemm_4x4_lib4.S` does not assemble correctly:

    kernel_dgemm_4x4_lib4.S: Assembler messages:
    kernel_dgemm_4x4_lib4.S:2053: Error: co-processor offset out of range
    kernel_dgemm_4x4_lib4.S:2122: Error: co-processor offset out of range
    kernel_dgemm_4x4_lib4.S:2163: Error: co-processor offset out of range
    kernel_dgemm_4x4_lib4.S:2247: Error: co-processor offset out of range
    kernel_dgemm_4x4_lib4.S:2302: Error: co-processor offset out of range
    kernel_dgemm_4x4_lib4.S:2302: Error: co-processor offset out of range
    kernel_dgemm_4x4_lib4.S:2302: Error: co-processor offset out of range
    kernel_dgemm_4x4_lib4.S:2302: Error: co-processor offset out of range
    kernel_dgemm_4x4_lib4.S:2302: Error: co-processor offset out of range
    kernel_dgemm_4x4_lib4.S:2370: Error: co-processor offset out of range
    kernel_dgemm_4x4_lib4.S:2530: Error: co-processor offset out of range
    kernel_dgemm_4x4_lib4.S:2621: Error: co-processor offset out of range
    kernel_dgemm_4x4_lib4.S:2621: Error: co-processor offset out of range
    kernel_dgemm_4x4_lib4.S:2621: Error: co-processor offset out of range
    kernel_dgemm_4x4_lib4.S:2621: Error: co-processor offset out of range
    kernel_dgemm_4x4_lib4.S:2621: Error: co-processor offset out of range

This is related to the `.LC00` and `.LC01` labels that hold zero constants that are used in various places as second parameter for the `fldd` instructions. The labels are not allowed to have too great offsets from their usages but apparently the assembler thinks this is the case in the above cases. I could not find a fix for that but my assembler foo is lacking ;)

BTW: there is another breakage on armv7a regarding bad instructions but I have a fix for that and will open a pull request soon.

The preprocessed output of `/usr/lib/gcc-cross/arm-linux-gnueabihf/8/cc1 -E -lang-asm -quiet -v -imultilib . -imultiarch arm-linux-gnueabihf -D MACRO_LEVEL=1 -D OS_LINUX -D TARGET_ARMV7A_ARM_CORTEX_A15 kernel_dgemm_4x4_lib4.S -mfpu=neon-vfpv4 -mfloat-abi=hard -mthumb -mtls-dialect=gnu -march=armv7-a+neon-vfpv4 -fno-directives-only` is attached. It's then fed into `/usr/lib/gcc-cross/arm-linux-gnueabihf/8/../../../../arm-linux-gnueabihf/bin/as -v -march=armv7-a -mfloat-abi=hard -mfpu=neon-vfpv4 -meabi=5 -o kernel_dgemm_4x4_lib4.o` which produces the errors above. The flow is easily visible by executing `arm-linux-gnueabihf-gcc   -DMACRO_LEVEL=1 -DOS_LINUX -mfpu=neon-vfpv4 -DTARGET_ARMV7A_ARM_CORTEX_A15   -c -o kernel_dgemm_4x4_lib4.o kernel_dgemm_4x4_lib4.S -v` (note the `-v`) in `kernel/armv7a`.

[cc3nUHts.s.log](https://github.com/giaf/blasfeo/files/3744421/cc3nUHts.s.log)

**Comment by stefanct** [comment]
I was able to fix these issues for `MACRO_LEVEL`s 0 and 1, but it isn't pretty.

For the macros (as used at level 1) instead of using fixed labels for the zeroing that refer to locations before the macros I am using macro parameters to pass the label name into them. For each location of macro invocation I am creating a new constant label (thus ensuring it's near enough when actually assembling).

However, the macro parameters are of course useless when the routines are assembler functions. For these cases I have introduced new constants (that get defined before the function but only if functions are used, i.e. `MACRO_LEVEL` 0) that are getting applied (vs. the macro parameters) according to the value of `MACRO_LEVEL`.

Apart from the ugliness of my changes they also do not fix the actual problem: for `MACRO_LEVEL` 2 the problem returns... so clearly some better solution by someone more familiar with gas is required.
